Forge Command Refactor, Command Porting, Export Command, Fallbacks

Jishuna opened the pull request:
- Refactors forge commands to use a similar system to bukkit.
- Ports /otg map and /otg spawn to forge. (/otg map is currently laggy without the BukkitRunnable system)
- Fixes /otg spawn on bukkit not defaulting to the users current world.
- Adds /otg lookup to both platforms, allowing you to lookup a biome by ID or name.
- Adds /otg export to both platforms, allowing you to export a worldedit selection as a BO3.
- Adds a ModCheck and ModCheckNot option to BO3s.
- Adds /otg mods to forge.
- Adds a block fallback system.

Jishuna committed:
Merge pull request #1 from PG85/1.12-OTG+-V8

1.12 otg+ v8

Jishuna committed:
Forge command refactor.

Jishuna committed:
Forge command refactor.

Jishuna committed:
Tab Complete Fixes, /otg lookup on Bukkit

Jishuna committed:
Tab Complete Fixes, /otg lookup on Bukkit

Jishuna committed:
/otg export
- Allows you to export a worldedit selection as a BO3, supports branching, tile entities, and center block selection.

Jishuna committed:
/otg export
- Allows you to export a worldedit selection as a BO3, supports branching, tile entities, and center block selection.

Jishuna committed:
Fix /otg export
- Now works properly on forge
- Attempt to fix build.gradle for forge (Game currently will not run from dev environment as it attempts to load the outdated worldedit mod, not sure how to stop this.)

Jishuna committed:
Fix /otg export
- Now works properly on forge
- Attempt to fix build.gradle for forge (Game currently will not run from dev environment as it attempts to load the outdated worldedit mod, not sure how to stop this.)

Jishuna committed:
Mod Checks
 - ModCheck and ModCheckNot can be used in BO3's to check against loaded mods.
 - No effect on spigot.

Jishuna committed:
Mod Checks
 - ModCheck and ModCheckNot can be used in BO3's to check against loaded mods.
 - No effect on spigot.

PG85 commented:
Hey @Coll1234567, thanks this is awesome! Will review asap and see if I can include it for v8 release. Would delay the release (again) ofcourse, so will have to see.

Jishuna committed:
/otg mods
 - Shows a list of loaded mods and their modid for use in ModChecks.

Jishuna committed:
/otg mods
 - Shows a list of loaded mods and their modid for use in ModChecks.

Jishuna committed:
Formatting Changes

Jishuna committed:
Formatting Changes

Jishuna committed:
Block  Fallbacks
 - Presets can now include a Fallbacks.ini file with a list of BlockFallbacks.
 - If a block is not present at runtime, OTG will attempt to use a fallback.
 - If no fallback is found OTG will defualt to air.
 - Implimentation still requires plenty of testing, possible improvements can be made.

Jishuna committed:
Block  Fallbacks
 - Presets can now include a Fallbacks.ini file with a list of BlockFallbacks.
 - If a block is not present at runtime, OTG will attempt to use a fallback.
 - If no fallback is found OTG will defualt to air.
 - Implimentation still requires plenty of testing, possible improvements can be made.

Jishuna committed:
Add spawnFromSapling() to CustomObject to (hopefully) stop saplings from consuming everything in their path.

Jishuna committed:
Biome-specific fog.

Jishuna committed:
Biome-specific fog.

Jishuna committed:
Misc refactoring and cleanup.

Jishuna committed:
Misc refactoring and cleanup.

Jishuna committed:
Fixed a wrong import.

Jishuna committed:
Fixed a wrong import.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/CustomObjectCollection.java:
```diff
@@ -45,95 +45,100 @@ private CustomObject loadObject(File file, String worldName)
+
+                    boolean passedEnable = object.onEnable();
+                    boolean passedChecks = object.loadChecks();
+                    if (!passedEnable || !passedChecks)
```
Should loadChecks() be executed when onEnable has failed?

If passedEnable is only used in 1 place, no need to make a var for it.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/CustomObjectCollection.java:
```diff
@@ -153,257 +158,290 @@ void reloadCustomObjectFiles()
+
+        // Only check the GlobalObjects if the WorldObjects directory has
+        // already been
+        // searched
```
could you make your ide not auto-collapse code and comments like this for the future? If there's any files with only formatting changes, could you revert those?

complex if statements etc I get, good to spread those over multiple lines.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/CustomObjectCollection.java:
```diff
@@ -153,257 +158,290 @@ void reloadCustomObjectFiles()
+                    } else
+                    {
+                        if (fileInDir.getName().toLowerCase().endsWith(
+                                ".bo4data") || fileInDir.getName().toLowerCase().endsWith(
```
Multiple lines here is a good idea IDE, just not this way XD

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bo2/BO2.java:
```diff
@@ -74,11 +75,15 @@ public boolean spawnForced(LocalWorld world, Random random, Rotation rotation, i
         // Spawn
         for (ObjectCoordinate point : data)
         {
+            if (point.material instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bo2/BO2.java:
```diff
@@ -381,20 +392,25 @@ private void readCoordinates()
+            }
+        }
+        
+        if (material instanceof RawMaterialData) {
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bo3/ObjectExtrusionHelper.java:
```diff
@@ -94,15 +94,15 @@ void extrude(LocalWorld world, Random random, int x, int y, int z)
                      --yi)
                 {
-                	world.setBlock(x + block.x, yi, z + block.z, block.material, block.metaDataTag, false);
+                	world.setBlock(x + block.x, yi, z + block.z, block.parseForWorld(world), block.metaDataTag, false);
```
refactor parseForWorld -> parseMaterial(LocalWorld world)?

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bo3/bo3function/BO3RandomBlockFunction.java:
```diff
@@ -155,6 +156,10 @@ public void spawn(LocalWorld world, Random random, int x, int y, int z, boolean
         {
             if (random.nextInt(100) < blockChances[i])
             {
+                if (blocks[i] instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bo3/checks/BlockCheck.java:
```diff
@@ -15,6 +15,7 @@
     @Override
     public boolean preventsSpawn(LocalWorld world, int x, int y, int z)
     {
+        toCheck.parseForWorld(world);
```
pls refactor toCheck to materialSet for me and parseForWorld to parseMaterials, so materialSet.parseMaterials, otherwise its not obvious what's going on.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bo3/BO3Config.java:
```diff
@@ -18,678 +24,697 @@
+		writer.comment(
+				"Also, use this only to store IDs/object names etc for your mod, DO NOT include things like character dialogue,");
+		writer.comment(
+				"messages on signs, loot lists etc 
```
Could define evaluate on the BO3Check level and implement for each child type? saves the instanceof checks.

PG85 commented on common/src/main/java/com/pg85/otg/configuration/world/WorldConfig.java:
```diff
@@ -382,6 +395,44 @@ public static DefaulWorldData createDefaultOTGWorldConfig(File settingsDir, Stri
+                }
+            }
+        }
+        return OTG.getEngine().toLocalMaterialData(DefaultMaterial.AIR, 0);
```
Have to test if never returning null won't cause problems, some callers may expect AIR and null to be distinct cases. A material that can't be parsed and doesn't have a fallback should return null, the caller should handle that case. If there's a lot of places that should default to AIR, create a parseFallBackOrAir method (or a theMethodThatCallsParseFallbackOrAir() method) and make those places use that method instead.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bo4/bo4function/BO4RandomBlockFunction.java:
```diff
@@ -170,6 +171,10 @@ public void spawn(LocalWorld world, Random random, int x, int y, int z, boolean
         {
             if (random.nextInt(100) < blockChances[i])
             {
+                if (blocks[i] instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bofunctions/BlockFunction.java:
```diff
@@ -44,6 +45,14 @@ public void load(List<String> args) throws InvalidConfigException
     }
+    
+    public LocalMaterialData parseForWorld(LocalWorld world) {
+        if (material instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/customobjects/bofunctions/BlockFunction.java:
```diff
@@ -71,13 +80,18 @@ public String makeString()
      */
     public void spawn(LocalWorld world, Random random, int x, int y, int z, boolean isOTGPlus)
     {
+        if (material instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/generator/ChunkProviderOTG.java:
```diff
@@ -304,7 +304,7 @@ private boolean addBiomeBlocksAndCheckWater(ChunkBuffer chunkBuffer)
                 final BiomeConfig biomeConfig = toBiomeConfig(this.biomeArray[(x + z * CHUNK_X_SIZE)]);
 
-                biomeConfig.surfaceAndGroundControl.spawn(this.localWorld.getSeed(), generatingChunk, chunkBuffer, biomeConfig, chunkCoord.getBlockX() + x, chunkCoord.getBlockZ() + z);
+                biomeConfig.surfaceAndGroundControl.spawn(localWorld, generatingChunk, chunkBuffer, biomeConfig, chunkCoord.getBlockX() + x, chunkCoord.getBlockZ() + z);
```
Currently fields and inner variables both use lowercase, so I prefer always using this. for fields to be clear. Should probably use different casing or underscores for fields / inner variables so you can't build when you forget a this.

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/GrassGen.java:
```diff
@@ -148,6 +148,7 @@ private void spawnGrouped(LocalWorld world, Random random, ChunkCoordinate chunk
     @Override
     protected void spawnInChunk(LocalWorld world, Random random, boolean villageInChunk, ChunkCoordinate chunkCoord)
     {
+        sourceBlocks.parseForWorld(world);
```
refactor to parseMaterial(Localworld world) / parseMaterials(Localworld world).

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/PlantType.java:
```diff
@@ -150,12 +155,36 @@ public String getName()
+    
+    private void parseMaterials(LocalWorld world, LocalMaterialData material, MaterialSet sourceBlocks)
+    {
+        if (material instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/PlantType.java:
```diff
@@ -150,12 +155,36 @@ public String getName()
+
+        if (sourceBlocks != null)
+        {
+            sourceBlocks.parseForWorld(world);
```
refactor to parseMaterial(LocalWorld world) / parseMaterials(LocalWorld world)

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/Resource.java:
```diff
@@ -97,6 +99,19 @@ public LocalMaterialData getMaterial()
+    
+    protected void parseMaterials(LocalWorld world, LocalMaterialData material, MaterialSet sourceBlocks)
+    {
+        if (material instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/Resource.java:
```diff
@@ -97,6 +99,19 @@ public LocalMaterialData getMaterial()
+
+        if (sourceBlocks != null)
+        {
+            sourceBlocks.parseForWorld(world);
```
refactor to parseMaterial(LocalWorld world) / parseMaterials(LocalWorld world)

PG85 commented on common/src/main/java/com/pg85/otg/generator/surface/MesaSurfaceGenerator.java:
```diff
@@ -333,6 +342,16 @@ else if (k1 > 0)
+
+    private LocalMaterialData checkAndParseRawData(LocalWorld world, LocalMaterialData data)
+    {
+        if (data instanceof RawMaterialData) {
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/generator/surface/SimpleSurfaceGenerator.java:
```diff
@@ -29,14 +30,21 @@ public LocalMaterialData getCustomBlockData(LocalWorld world, BiomeConfig biomeC
-    protected final void spawnColumn(LocalMaterialData defaultSurfaceBlock, LocalMaterialData defaultGroundBlock, GeneratingChunk generatingChunk, ChunkBuffer chunkBuffer, BiomeConfig biomeConfig, int x, int z)
+    protected final void spawnColumn(LocalWorld world, LocalMaterialData defaultSurfaceBlock, LocalMaterialData defaultGroundBlock, GeneratingChunk generatingChunk, ChunkBuffer chunkBuffer, BiomeConfig biomeConfig, int x, int z)
     {
+        if (defaultSurfaceBlock instanceof RawMaterialData) {
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/generator/surface/SimpleSurfaceGenerator.java:
```diff
@@ -124,6 +138,15 @@ else if (surfaceBlocksCount > 0)
 
+    private LocalMaterialData checkAndParseRawData(LocalWorld world, LocalMaterialData data)
+    {
+        if (data instanceof RawMaterialData) {
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return itself. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/util/helpers/MaterialHelper.java:
```diff
@@ -54,8 +56,7 @@ else if(CachedMaterials.containsKey(name))
     	{
-    		CachedMaterials.put(originalName, null);
-    		throw ex;
+    		material = new RawMaterialData(name);
```
invalidconfigexceptions aren't implemented consistently atm to properly report and bubble up any exceptions related to incorrect user input. Ideally they should be used and callers should catch and handle them. I think readMaterial isn't just used atm to read material names from user-made configs though, so invalidconfigexceptions should probably be reimplemented at the config level, not the readmaterial level (readmaterial can just return null).

Add a // TODO with a note?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/SmallLakeGen.java:
```diff
@@ -18,6 +18,7 @@
     private final boolean[] BooleanBuffer = new boolean[2048];
     private int maxAltitude;
     private int minAltitude;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/SmallLakeGen.java:
```diff
@@ -100,6 +101,11 @@ public void spawn(LocalWorld world, Random rand, boolean villageInChunk, int x,
         // y = floor
         y -= 4;
+        
+        if (!bLoaded  ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/SurfacePatchGen.java:
```diff
@@ -24,6 +24,7 @@
     private final NoiseGeneratorSurfacePatchOctaves noiseGen;
     private final Random random;
     private final MaterialSet sourceBlocks;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/UnderWaterOreGen.java:
```diff
@@ -13,6 +13,7 @@
 {
     private final int size;
     private final MaterialSet sourceBlocks;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/VeinGen.java:
```diff
@@ -23,6 +23,7 @@
     int oreSize; // Average size of a ore in the vein
     MaterialSet sourceBlocks; // Blocks for the ore to spawn in
     private double veinRarity; // Chance for the vein to spawn in a chunk
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/WellGen.java:
```diff
@@ -17,6 +17,7 @@
     private final LocalMaterialData slab;
     private final LocalMaterialData water;
     private final MaterialSet sourceBlocks;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/SurfacePatchGen.java:
```diff
@@ -117,6 +118,11 @@ public void spawn(LocalWorld world, Random rand, boolean villageInChunk, int x,
         if (y < minAltitude || y > maxAltitude)
             return;
+        
+        if (!bLoaded  ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/UnderWaterOreGen.java:
```diff
@@ -72,6 +73,11 @@ public void spawn(LocalWorld world, Random rand, boolean villageInChunk, int x,
             return;
         }
+        
+        if (!bLoaded  ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/VeinGen.java:
```diff
@@ -133,6 +134,11 @@ protected void spawnInChunk(LocalWorld world, Random random, boolean villageInCh
         // Find all veins that reach this chunk, and spawn them
         int searchRadius = (this.maxRadius + 15) / 16;
+        
+        if (!bLoaded  ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/WellGen.java:
```diff
@@ -95,6 +96,12 @@ public void spawn(LocalWorld world, Random random, boolean villageInChunk, int x
             return;
         }
+        
+        if (!bLoaded  ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/util/materials/MaterialSetEntry.java:
```diff
@@ -44,6 +46,14 @@ public int hashCode()
 
+    public void parseForWorld(LocalWorld world)
+    {
+        if (material instanceof RawMaterialData)
```
create LocalMaterialData.parseMaterial(LocalWorld) and call that here instead, add this code to RawMaterialData.parseMaterial, for ForgeMaterialData/BukkitMaterialData, make parseMaterial return. Saves the instanceof call for already parsed materials. Its a good question tho if using inheritance that way isn't slower, assuming it isn't.. Also good to keep this logic in one place.

PG85 commented on common/src/main/java/com/pg85/otg/util/materials/MaterialSetEntry.java:
```diff
@@ -44,6 +46,14 @@ public int hashCode()
         }
     }
 
+    public void parseForWorld(LocalWorld world)
```
refactor to parseMaterial(LocalWorld world)

PG85 commented on platforms/bukkit/build.gradle:
```diff
@@ -19,12 +20,17 @@ repositories
 {
     compile 'org.bukkit:bukkit:' + project.ext.bukkitVersion
     compile 'org.spigotmc:spigot:' + project.ext.bukkitVersion
+    compileOnly 'com.sk89q.worldedit:worldedit-bukkit:' + project.ext.worldeditVersion
```
Would be interested to know what happens if you import and use WE classes at mod init, and then run the client without WE. It should crash with something like a NoClassDefFoundError / ClassNotFoundException / MethodNotFoundException. It think it doesn't currently crash because the OTG classes that use the WE dependencies are never loaded if WE isnt installed, OTG commands never reach that code, which is nice.

PG85 commented on platforms/bukkit/src/main/java/com/pg85/otg/bukkit/commands/ExportCommand.java:
```diff
@@ -0,0 +1,85 @@
+package com.pg85.otg.bukkit.commands;
```
Extremely sweet! Will have to test this. Expect ppl to start asking for improvements immediately :p

PG85 commented on platforms/bukkit/src/main/java/com/pg85/otg/bukkit/commands/SpawnCommand.java:
```diff
@@ -1,25 +1,25 @@
 package com.pg85.otg.bukkit.commands;
```
Awesome, ofc you realise that the first thing ppl will want is an undo function and a holo-BO3 they can move around before placing :P.

PG85 commented on platforms/bukkit/src/main/java/com/pg85/otg/bukkit/util/BO3Creator.java:
```diff
@@ -0,0 +1,343 @@
+
+                    if (branch)
+                    {
+                        chunkCoordinates = new ChunkCoordIntPair((int) Math.floor(
```
Really need to find a better name than ChunkCoordIntPair ><

PG85 commented on platforms/bukkit/src/main/java/com/pg85/otg/bukkit/util/BO3Creator.java:
```diff
@@ -0,0 +1,343 @@
+package com.pg85.otg.bukkit.util;
```
This is going to need some heavy testing I imagine :o

PG85 commented on platforms/bukkit/src/main/java/com/pg85/otg/bukkit/util/NBTHelper.java:
```diff
@@ -1,300 +1,331 @@
 package com.pg85.otg.bukkit.util;
 
+import java.lang.reflect.Field;
```
If there's any files where only the formatting was changed, can you revert those?

PG85 commented on platforms/bukkit/src/main/resources/plugin.yml:
```diff
@@ -52,4 +52,10 @@ permissions:
         default: op
-        description: Teleport to the given biome (name or id, max range 1000 chunks)
\ No newline at end of file
+        description: Teleport to the given biome (name or id, max range 1000 chunks)
```
I thought it was 10k? Could be wrong..

PG85 commented on common/src/main/java/com/pg85/otg/configuration/biome/BiomeConfig.java:
```diff
@@ -696,6 +708,28 @@ protected void writeConfigSettings(SettingsMap writer)
+        writer.putSetting(BiomeStandardValues.FOG_COLOR, this.fogColor, "Biome fog color.");
+
+        writer.putSetting(BiomeStandardValues.FOG_DENSITY, this.fogDensity,
+                "How dense the fog is this biome is, Float value from 0.0 to 1.0.",
```
is this is

PG85 commented on common/src/main/java/com/pg85/otg/configuration/biome/BiomeConfig.java:
```diff
@@ -696,6 +708,28 @@ protected void writeConfigSettings(SettingsMap writer)
+
+        writer.putSetting(BiomeStandardValues.FOG_DENSITY, this.fogDensity,
+                "How dense the fog is this biome is, Float value from 0.0 to 1.0.",
+                "A value of 0 produces almost no fog while a value of 1 will cover the entire screen with fog.");
```
A value of 0 produces almost no fog while a value of 1 will cover the entire screen with fog.
A value of 0 produces almost no fog, a value of 1 covers the screen with fog.

3 words less, didn't lose any meaning :). They always tell me to make things shorter, just a writing tip ><

PG85 commented on common/src/main/java/com/pg85/otg/configuration/biome/BiomeConfig.java:
```diff
@@ -696,6 +708,28 @@ protected void writeConfigSettings(SettingsMap writer)
+                "A value of 0 produces almost no fog while a value of 1 will cover the entire screen with fog.");
+
+        writer.putSetting(BiomeStandardValues.FOG_TIME_WEIGHT, this.fogTimeWeight,
+                "How much the world time should affect the fog color, Float value from 0.0 to 1.0.",
```
Normies don't know what a float is, maybe use decimal number.

PG85 commented on platforms/forge/build.gradle:
```diff
@@ -33,9 +34,27 @@ minecraft {
 dependencies
 {
     compile project(':common')	
+	provided ('com.sk89q.worldedit:worldedit-forge-mc1.11:6.1.6') {
```
Would be interested to know what happens if you import and use WE classes at mod init, and then run the client without WE. It should crash with something like a NoClassDefFoundError / ClassNotFoundException / MethodNotFoundException. It think it doesn't currently crash because the OTG classes that use the WE dependencies are never loaded if WE isnt installed, OTG commands never reach that code, which is nice.

PG85 commented on platforms/forge/src/main/java/com/pg85/otg/forge/commands/OTGCommandHandler.java:
```diff
@@ -0,0 +1,138 @@
+package com.pg85.otg.forge.commands;
```
Very nice refactoring of Forge commands, also great that Forge<>Spigot are more aligned now!

PG85 commented on platforms/forge/src/main/java/com/pg85/otg/forge/commands/runnables/MapWriter.java:
```diff
@@ -0,0 +1,233 @@
+    {
+        // OTG.log(LogMarker.DEBUG, "BukkitWorld::UUID:: {}",
+        // world.getWorldInfo().geti;
+        LocalWorld bukkitWorld = OTG.getWorld(WorldHelper.getName(world2));
```
forgeWorld?

PG85 commented on platforms/forge/src/main/java/com/pg85/otg/forge/commands/runnables/MapWriter.java:
```diff
@@ -0,0 +1,233 @@
+package com.pg85.otg.forge.commands.runnables;
```
Elegant solution! If this is executed asynchronously though, it may access shared resources in an unsafe way, which could cause problems f.e. if the world is unloaded while the MapWriter is running?

PG85 commented on platforms/forge/src/main/java/com/pg85/otg/forge/events/client/ClientFogHandler.java:
```diff
@@ -0,0 +1,307 @@
+package com.pg85.otg.forge.events.client;
```
Awesome! Where'd you get the info/inspiration/code from to write this? Any other mods or MC classes you looked at? Might be nice to add a reference (if any).

PG85 commented on platforms/forge/src/main/java/com/pg85/otg/forge/util/BO3Creator.java:
```diff
@@ -0,0 +1,338 @@
+package com.pg85.otg.forge.util;
```
Very nice, this'll need a shitton of testing :o.

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/ReedGen.java:
```diff
@@ -15,6 +15,7 @@
     private final int maxAltitude;
     private final int minAltitude;
     private final MaterialSet sourceBlocks;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/PlantType.java:
```diff
@@ -150,12 +155,36 @@ public String getName()
     void spawn(LocalWorld world, int x, int y, int z)
     {
+        
+        if (!bLoadedTop) {
```
Why use 2 booleans if you're loading both materials in sequence anyway?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/OreGen.java:
```diff
@@ -17,6 +17,7 @@
     private final int maxSize;
     private final int minAltitude;
     private final MaterialSet sourceBlocks;
+    private boolean bLoaded = true;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/LiquidGen.java:
```diff
@@ -16,6 +16,7 @@
     private final int maxAltitude;
     private final int minAltitude;
     private final MaterialSet sourceBlocks;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/CactusGen.java:
```diff
@@ -37,6 +38,11 @@ public CactusGen(BiomeConfig biomeConfig, List<String> args) throws InvalidConfi
     {
         int y = RandomHelper.numberInRange(rand, minAltitude, maxAltitude);
+        
+        if (!bLoaded) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/BoulderGen.java:
```diff
@@ -15,6 +15,7 @@
     private MaterialSet sourceBlocks;
     private int minAltitude;
     private int maxAltitude;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/CactusGen.java:
```diff
@@ -17,6 +17,7 @@
     private int minAltitude;
     private int maxAltitude;
     private MaterialSet sourceBlocks;
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/AboveWaterGen.java:
```diff
@@ -9,6 +9,8 @@
 
 public class AboveWaterGen extends Resource
 {
+    private boolean bLoaded = false;
```
Could put this in Resource instead?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/ReedGen.java:
```diff
@@ -99,6 +100,11 @@ public void spawn(LocalWorld world, Random rand, boolean villageInChunk, int x,
             return;
         }
+        
+        if (!bLoaded  ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/OreGen.java:
```diff
@@ -78,6 +79,12 @@ public String toString()
     public void spawn(LocalWorld world, Random rand, boolean villageInChunk, int x, int z)
     {
+        
+        if (!bLoaded ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/LiquidGen.java:
```diff
@@ -85,6 +86,11 @@ public void spawn(LocalWorld world, Random rand, boolean villageInChunk, int x,
         if (!world.isNullOrAir(x, y, z, false) && (!sourceBlocks.contains(world.getMaterial(x, y, z, false))))
             return;
+        
+        if (!bLoaded ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/BoulderGen.java:
```diff
@@ -38,6 +39,11 @@ public void spawn(LocalWorld world, Random random, boolean villageInChunk, int x
             return;
         }
+        
+        if (!bLoaded ) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

PG85 commented on common/src/main/java/com/pg85/otg/generator/resource/AboveWaterGen.java:
```diff
@@ -26,13 +28,19 @@ public void spawn(LocalWorld world, Random rand, boolean villageInChunk, int x,
         if (y == -1)
             return;
 
+       if (!bLoaded) {
```
Could put this in a method in Resource instead, no repeating code for each Gen class?

Jishuna committed:
Refactoring to remove RawMaterialData.
 - Now handled by BukkitMaterialData and ForgeMaterialData

Jishuna committed:
Refactoring to remove RawMaterialData.
 - Now handled by BukkitMaterialData and ForgeMaterialData

Jishuna committed:
Fixed a crash with BO4s.

Jishuna committed:
Fixed a crash with BO4s.

Jishuna committed:
More misc fixes.

Jishuna committed:
More misc fixes.

Jishuna committed:
Fix /otg map on forge.

Jishuna committed:
Fix /otg map on forge.

Jishuna committed:
Fixes for /otg export.

Jishuna committed:
Fixes for /otg export.

PG85 committed:
Merge pull request #335 from Coll1234567/1.12-OTG+-V8-Bugs

Add spawnFromSapling() to CustomObject

PG85 committed:
BO3 customstructures fix (breadslicing)

PG85 committed:
r7 version update / minor refactor

PG85 committed:
r7 version number

PG85 committed:
v8_r7 forge modded cavegen fix

PG85 committed:
v8_r7: Disable OTG-Core when OTG not present

PG85 committed:
Merge branch '1.12-OTG+-V8' of https://github.com/PG85/OpenTerrainGenerator into 1.12-OTG+-V8

PG85 committed:
v8_r8: UI fixes

- Fixed the presets menu, cloning wasn't possible for presets that were in use, also no preset was selected when entering the menu, causing MC to crash when pressing Clone.

PG85 committed:
v8_r8: /otg tp spigot, modpack config fixes, GUI fixes, portals/tp not underground

- Fixed spaced in /otg tp biome names not working for spigot, fe "Webbed Forest".
- Fixed /otg tp and portals mostly tp'ing underground.
- Removed some logging.
- Fixed crashes related to modpack configs.
- Fixed crashes related to O menu.

PG85 committed:
v8.1_r1: fixed nullpointer FrozenSurfaceHelper/SaplingListener, fixed mac/linux filepaths

- Fixes a nullpointerexception with FrozenSurfaceHelper (https://github.com/PG85/OpenTerrainGenerator/issues/348)
- Fixes a nullpointerexception with SaplingListener (https://github.com/PG85/OpenTerrainGenerator/issues/361)
- Fixes hopefully the last file path problem for mac/unix.
- Minor cleanup.

PG85 committed:
v8.1_r1: Update version number

PG85 committed:
v8.2_r1: BoP/RTG o menu fix, Galacticraft planet switching fix, snow on modded blocks fix

- Fixes problems with BoP/RTG, BoP/RTG worlds that were created before
 OTG was installed had problems with the "o" menu (https://github.com/PG85/OpenTerrainGenerator/issues/345).
- Fixes a crash with Galacticraft when switching planets (https://github.com/PG85/OpenTerrainGenerator/issues/324).
- Snow is no longer placed on non-solid modded block (plants etc) (https://github.com/PG85/OpenTerrainGenerator/issues/362).

PG85 committed:
8.3_r1: otg+ multiple customstructures fix

- Fixed a bug that allowed only one customstructure type per biome when using multiple Customstructure() resources.

PG85 committed:
v8.3_r3: Pregen fixes

This should solve pregenerator problems:
- for non-OTG+ worlds, should solve missing chunks/branches.
- for OTG+ worlds, should solve compresscache crash. not sure if corrupted OTG+ worlds can be recovered tho.

PG85 committed:
v8.3_r4: Pregen, BO3Groups

- Made sure all pregenerated chunks are populated
- Fixed BO3Groups

PG85 committed:
v8.3_r4: Bo4data compression

- Added compression for bo4data files. Files much smaller, startup times quicker, much less stutter when exploring. Doesn't seem to affect pregen speed/memory usage much tho. Ppl will need to re-generate their bo4data files.

PG85 committed:
v8.3_r4: pregen backwards compat

- Added a fix that should hopefully fix corrupted pregendata files for older v8 versions.

PG85 committed:
v8.3: Update version number for release

PG85 committed:
Merge branch '1.12-OTG+-V8' of https://github.com/Coll1234567/OpenTerrainGenerator into 1.12-OTG+-V8
